Tune LitinskiTransformation pass

mtreinish opened the pull request:
<!--
⚠️  If you do not respect this template, your pull request will be closed.
⚠️  Your pull request title should be short detailed and understandable for all.
⚠️  Also, please add a release note file using reno if the change needs to be documented in the release notes.
⚠️  If your pull request fixes an open issue, please link to the issue. Use "Fixes #XXXX" if this PR *fully* closes the issue XXXX.  
☢️  If you used an AI tool to code this PR, add "AI tool used: <Name and version of the tool>". For example, "AI tool used: Microsoft Copilot Chat with GPT-5". Failing to disclose the use of AI tools may result in the PR being closed without further review.  


- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This commit optimizes the implementation of the LitinskiTransformation pass. The improvements fall into a couple of categories, but most of the runtime performance improvements will come from removing unnecessary allocations.

From a quick glance the Clifford struct looks like it could also use a similar once over, mostly to get it to leverage Qiskit's types more effectively if nothing else. But I kept the changes there minimal for this commit and only modified the method used by The LitinskiTransformation directly.

### Details and comments

mtreinish committed:
Tune LitinskiTransformation pass

This commit optimizes the implementation of the LitinskiTransformation
pass. The improvements fall into a couple of categories, but most of the
runtime performance improvements will come from removing unecessary
allocations.

From a quick glance the Clifford struct looks like it could also use a
similar once over, mostly to get it to leverage Qiskit's types more
effectively if nothing else. But I kept the changes there minimal for
this commit and only modified the method used by The
LitinskiTransformation directly.

mtreinish commented:
To benchmark this I used the benchmark @alexanderivrii described in the PR summary on #14753 with this script:

```python
import statistics
import time

from qiskit.converters import circuit_to_dag
from qiskit.circuit.library import QFTGate
from qiskit.transpiler.passes import LitinskiTransformation
from qiskit.transpiler import Target
from qiskit import transpile, QuantumCircuit

litinski = LitinskiTransformation()

i = 2
while i <= 1024:

    qft = QuantumCircuit(i)
    qft.append(QFTGate(i), qft.qubits)
    target = Target.from_configuration(["cx", "sx", "rz"], i)
    tqft = transpile(qft, target=target)
    dag = circuit_to_dag(tqft, False)
    times = []
    for _ in range(10):
        start = time.perf_counter()
        litinski.run(dag)
        stop = time.perf_counter()
        times.append(stop - start)
    print(f"nqubits: {i}, runtime: {statistics.geometric_mean(times)}")
    i = i * 2
```

It yielded these results on my local system:
<img width="5760" height="4320" alt="litinski" src="https://github.com/user-attachments/assets/a31df8ba-f16b-40f6-940a-caef0d5be6d0" />

alexanderivrii commented:
Thanks @mtreinish! I have just now run your script for the most problematic case of `QFTGate(1024)` (with the only addition of also printing the runtime for each individual run) and I see no significant difference between the main branch, #15217 (which extends the support to `PauliProductMeasurement` instructions and also refactors the code a bit, though this refactoring is orthogonal to your changes), and this PR. For each of the 3 code versions it takes about 43 seconds to run the transform. For reference, this is done on my mac laptop and python 3.13.

In any case, for the interest of helping @Cryoris to prepare for QDC, I would suggest merging #15217 first, and then merging your PR.

mtreinish committed:
Merge remote-tracking branch 'origin/main' into tune-litinski

Besides adapting to the rebase this also improves the SparseObservable
construction.

mtreinish committed:
Remove unnecessary Vec for no clbits

mtreinish committed:
Merge branch 'main' into tune-litinski

alexanderivrii commented:
I have locally tried ASV benchmarks in #15386 evaluating this PR on top of #15315, and indeed see a nice additional speedup:
```
Benchmarks that have improved:

| Change   | Before [ceb43c48] <speedup_litinski>   | After [a464d175] <speedup_and_tune>   |   Ratio | Benchmark (Parameter)                                                                           |
|----------|----------------------------------------|---------------------------------------|---------|-------------------------------------------------------------------------------------------------|
| -        | 733±9μs                                | 662±2μs                               |    0.9  | passes.LitinskiTransformationPassQFTBenchmarks.time_litinski_transformation_no_fix_clifford(10) |
| -        | 183±1μs                                | 164±0.7μs                             |    0.9  | passes.LitinskiTransformationPassQFTBenchmarks.time_litinski_transformation_no_fix_clifford(5)  |
| -        | 17.3±0.5ms                             | 15.5±0.3ms                            |    0.9  | passes.LitinskiTransformationPassQasmBenchmarks.time_square_heisenberg                          |
| -        | 30.1±2ms                               | 26.4±0.5ms                            |    0.88 | passes.LitinskiTransformationPassQFTBenchmarks.time_litinski_transformation_fix_clifford(100)   |
| -        | 780±6μs                                | 672±9μs                               |    0.86 | passes.LitinskiTransformationPassQFTBenchmarks.time_litinski_transformation_fix_clifford(10)    |
| -        | 204±2μs                                | 170±5μs                               |    0.83 | passes.LitinskiTransformationPassQFTBenchmarks.time_litinski_transformation_fix_clifford(5)     |

Benchmarks that have stayed the same:

| Change   | Before [ceb43c48] <speedup_litinski>   | After [a464d175] <speedup_and_tune>   |   Ratio | Benchmark (Parameter)                                                                             |
|----------|----------------------------------------|---------------------------------------|---------|---------------------------------------------------------------------------------------------------|
|          | 816±2ms                                | 806±2ms                               |    0.99 | passes.LitinskiTransformationPassQFTBenchmarks.time_litinski_transformation_no_fix_clifford(1000) |
|          | 121±20μs                               | 120±10μs                              |    0.99 | passes.LitinskiTransformationPassQasmBenchmarks.time_eoh                                          |
|          | 835±4ms                                | 814±2ms                               |    0.98 | passes.LitinskiTransformationPassQFTBenchmarks.time_litinski_transformation_fix_clifford(1000)    |
|          | 28.1±0.07ms                            | 26.0±0.1ms                            |    0.92 | passes.LitinskiTransformationPassQFTBenchmarks.time_litinski_transformation_no_fix_clifford(100)  |
|          | 316±1ms                                | 292±2ms                               |    0.92 | passes.LitinskiTransformationPassQasmBenchmarks.time_dtc                                          |
|          | 13.7±0.1ms                             | 12.6±0.07ms                           |    0.91 | passes.LitinskiTransformationPassQasmBenchmarks.time_qaoa                                         |
|          | 28.8±0.4ms                             | 26.3±0.1ms                            |    0.91 | passes.LitinskiTransformationPassQasmBenchmarks.time_qft                                          |

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

alexanderivrii reviewed (commented):
Thanks for doing this Matthew. Do you know where the improvements come from: using `BitTerm` instead of `(z, x)`, using the builder interface, or something else?

mtreinish committed:
Merge remote-tracking branch 'origin/main' into tune-litinski

mtreinish committed:
Merge branch 'main' into tune-litinski

mtreinish commented:
> Thanks for doing this Matthew. Do you know where the improvements come from: using `BitTerm` instead of `(z, x)`, using the builder interface, or something else?

I don't have detailed profiling to know exactly where the improvements are coming from. But I expect most of the runtime improvement is from removing the unnecessary allocations things like avoiding collecting indices into a new vec for every instruction we push on the dag. The `BitTerm` usage here falls into this category because it removes at least one vec allocation (I think it might be two or three but I'd have to trace through the sparse observable code again to be sure) as part of creating a pauli evolution gate. The other changes like using the builder interface or moving to matching on `StandardGate` and `StandardInstruction` instead of string names I expect have a smaller impact but still should speed things up a little bit.

alexanderivrii commented on crates/transpiler/src/passes/litinski_transformation.rs:
```diff
@@ -77,11 +73,22 @@ pub fn run_litinski_transformation(
+            }
+        })
+        .sum();
+    let clifford_count = dag.size(false)? - rotation_count;
```
Technically, `clifford_count` also includes the number of measurements, but it probably does not really matter as this is only used in `Vec::with_capacity`.

mtreinish replied:
It's just bad variable naming as this is leftover from when I first opened this pr and we didn't have ppm support here yet. `rotation_count` includes the number of measurements too, its the sum of the instruction counts from the instructions in `HANDLED_INSTRUCTION_NAMES`, which includes `"measure"`.

mtreinish replied:
I renamed the variable in https://github.com/Qiskit/qiskit/pull/15315/commits/371532369575c1e6c4a17fd6d0b819cd2cd4f152 to make this clearer.

alexanderivrii commented on crates/transpiler/src/passes/litinski_transformation.rs:
```diff
@@ -182,29 +280,48 @@ pub fn run_litinski_transformation(
+                        clifford.get_inverse_z(dag.get_qargs(inst.qubits)[0].index());
+                    let mut z: Vec<bool> = Vec::with_capacity(terms.len());
+                    let x: Vec<bool> = terms
+                        .into_iter()
```
Now `z` and `x` are treated a bit differently. Do you think that this is faster than also preallocating `x` with the given capacity and calling `terms.iter().for_each() { ... }`?

mtreinish replied:
It's probably about the same performance-wise. I wouldn't bother with a iterator chain in that case though and just use a for loop.

Although this is a good point that I've introduced an extra allocation for the measurement path which could be avoided. The only way I can think to avoid it is by adding another method which returns the split x a nd z vecs instead of a bit term vec which is similar to the function before the modifications in this pr. I'll play with it later and see if there is a path to avoid a ton of duplication but minimize the allocations for measurements.

mtreinish replied:
I did this in: https://github.com/Qiskit/qiskit/pull/15315/commits/a76941bf32307e108a5400905726402dbd76776b

alexanderivrii reviewed (commented):
Thanks Matthew for doing this, I have left two minor questions but would be happy to merge this without any changes.

mtreinish committed:
Use a more clear variable name for computing clifford count

mtreinish committed:
Merge remote-tracking branch 'origin/main' into tune-litinski

mtreinish committed:
Add separate method for PPM inverse z

The earlier changes in this PR branch had modified the
`Clifford::get_inverse_z()` method to be optimized for building a sparse
observable. This however came at the cost of extra overhead for the
pauli product measurement path. The overhead came in two ways, it resulted
in an extra vec allocation to collect the bit terms, but also in an extra
iteration over all the terms to recompute the z and x components for the
pauli product measurement. This was mainly a bias left from the original
version of this PR which was written before PauliProductMeasurement was
introduced or used in the pass. This commit fixes this oversight by
adding a second method to Clifford for PPM usage of get_inverse_z that
returns the data in a similar format to what was there before seperate
pauli z and x Vecs.

alexanderivrii reviewed (approved):
Thanks! 

Possibly in future it makes sense to port the `Pauli` class to rust (being internally represented as `Vec<BitTerm>` or something similar) and then change `PauliProductMeasurement` to keep just a `pauli` and a `sign`. Then we wouldn't need both `inverse_z` methods.
